refactor(profile): remove Profile(), loadProfile(), createProfile() singleton dependency

Kribylet opened the pull request:
This PR aims to simplify code in Profile and to remove many dependencies on Settings being a singleton.

If this PR is accepted a Profile object can be constructed without reference to Settings. Instead, relevant information is fetched during Core instantiation. This is a step towards completely decoupling Profile from Core and Settings.

renamed "name" method arguments in Profile to explicitly avoid shadowing class variable
made Settings reference const in Settings::createPersonal (no code changes)
extracted error logic from loadProfile and createProfile to increase readability.
moved Core initialization to loadProfile/createProfile
made loadProfile always require password QString (avoid name, settings, password) signature
made newly created profiles save an initial .tox file immediately.

Tested:
Creating profile w/o password
Creating profile with password
Loading profile
Creating existing profile error
Loading profile in use error
Loading passworded profile w/o password error
Loading passworded profile with wrong password error

General testing:
Adding friend
Chatting with friend
Creating group chat
Chatting in group chat
Logging in and out after adding friend/groupchat

- [x] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/5729)
<!-- Reviewable:end -->

Kribylet committed:
refactor(profile): use const reference in createNew/loadProfile

Kribylet committed:
refactor(profile): use const name reference in Profile()

Kribylet committed:
refactor(settings): declare createPersonal const

Kribylet committed:
refactor(profile): extract loadProfile error logic

Kribylet committed:
refactor(profile): simplify Profile constructor

Simplifies Profile constructor by passing Settings instance by
reference during loadProfile and createNewProfile calls.

Kribylet committed:
refactor(Profile): clang-format affected files

Diadlo reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 4 of 4 files at r3, 2 of 2 files at r4, 2 of 2 files at r5, 4 of 4 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5729#-:-LiStgn9-REcw89nrry-:b-jbsf8z)* status: :shipit: complete! 1 of 1 LGTMs obtained

___
*[src/persistence/profile.h, line 41 at r4](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsAjs75uPOVlJ82_M:-LiSsAjs75uPOVlJ82_N:b1eun5f) ([raw file](https://github.com/qtox/qtox/blob/897f49f850abbd9d8eb536be2d14ee05b971cb02/src/persistence/profile.h#L41)):*
> ```objc
>     Q_OBJECT
> 
>     enum LoadToxDataError
> ```

`enum class`?
___
*[src/persistence/profile.h, line 121 at r4](https://reviewable.io/reviews/qtox/qtox/5729#-LiSrv3f4C65LLBeQ2VG:-LiSrv3g9dTaXlWU_VMx:b77rjlc) ([raw file](https://github.com/qtox/qtox/blob/897f49f850abbd9d8eb536be2d14ee05b971cb02/src/persistence/profile.h#L121)):*
> ```objc
>     static std::unique_ptr<ToxEncrypt> loadToxData(const QString& password, const QString& filePath,
>                                                    QByteArray& data, LoadToxDataError& error);
>     static bool logLoadToxDataError(const LoadToxDataError& error, const QString& path);
> ```

Used only `cpp`. Move it there in anonymous namespace. Same for `LoadToxDataError`
___
*[src/persistence/profile.h, line 53 at r5](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsY3X9ZNt7D6f9W-C:-LiSsY3YDMHqIo2RRNJ8:b-ya81ng) ([raw file](https://github.com/qtox/qtox/blob/5395cdb5b6b738bd8c01cc70000de4f86ae55b60/src/persistence/profile.h#L53)):*
> ```objc
>     };
> 
>     enum CreateToxDataError
> ```

Same as above
___
*[src/persistence/profile.h, line 134 at r5](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsZlv-sIf0bXDOQzV:-LiSsZlv-sIf0bXDOQzW:b-ya81ng) ([raw file](https://github.com/qtox/qtox/blob/5395cdb5b6b738bd8c01cc70000de4f86ae55b60/src/persistence/profile.h#L134)):*
> ```objc
>     static std::unique_ptr<ToxEncrypt> createToxData(const QString& name, const QString& password,
>                                                      CreateToxDataError& error);
>     static bool logCreateToxDataError(const CreateToxDataError& error, const QString& userName);
> ```

Same as above
___
*[src/persistence/profile.cpp, line 179 at r4](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsOwD0I5BZpnP58Ff:-LiSsOwD0I5BZpnP58Fg:b-nfxcax) ([raw file](https://github.com/qtox/qtox/blob/897f49f850abbd9d8eb536be2d14ee05b971cb02/src/persistence/profile.cpp#L179)):*
> ```cpp
>     Profile* p = new Profile(name, password, false, toxsave, std::move(tmpKey));
> 
>     return p;
> ```

Return inplace
___
*[src/persistence/profile.cpp, line 304 at r5](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsxqe1f-0FI4OxeOW:-LiSsxqe1f-0FI4OxeOX:bj60bym) ([raw file](https://github.com/qtox/qtox/blob/5395cdb5b6b738bd8c01cc70000de4f86ae55b60/src/persistence/profile.cpp#L304)):*
> ```cpp
> 
>     if (ProfileLocker::hasLock()) {
>         error = CREATE_TOX_DATA_PROFILE_LOCKED;
> ```

`goto end`

It can lead to problems if someone will add logic somewhere between `if`s.


<!-- Sent from Reviewable.io -->

Diadlo reviewed (commented):
<img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5729#-:-LiStmX-6KQqAzhgrMrN:b8b79qu)* status: 0 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

Kribylet committed:
fix(profile): write .tox file immediately on creation

Fixes issue where .tox file of a new profile wasn't written until
connection was established.

Kribylet commented:
Fixed enum class, moved indicated functionality to anonymous namespace, return in-place added in commit, though removed again in later commit as operations are needed between instantiation and return. Added goto end calls. Found and corrected small bug present in both master and PR where newly created profiles write .ini files but not .tox files immediately.

Kribylet reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5729)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/profile.h, line 41 at r4](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsAjs75uPOVlJ82_M:-Lk-qn9u8a47ZoIrr9AL:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/897f49f850abbd9d8eb536be2d14ee05b971cb02/src/persistence/profile.h#L41)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

`enum class`?
</blockquote></details>

Done.
___
*[src/persistence/profile.h, line 121 at r4](https://reviewable.io/reviews/qtox/qtox/5729#-LiSrv3f4C65LLBeQ2VG:-Lk-qp2G0oMY0j7_2UrD:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/897f49f850abbd9d8eb536be2d14ee05b971cb02/src/persistence/profile.h#L121)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Used only `cpp`. Move it there in anonymous namespace. Same for `LoadToxDataError`
</blockquote></details>

Done.
___
*[src/persistence/profile.h, line 53 at r5](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsY3X9ZNt7D6f9W-C:-Lk-qo4BAUMjiiUYBBLf:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5395cdb5b6b738bd8c01cc70000de4f86ae55b60/src/persistence/profile.h#L53)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Same as above
</blockquote></details>

Done.
___
*[src/persistence/profile.h, line 134 at r5](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsZlv-sIf0bXDOQzV:-Lk-qpHfCI0H5gCSAWym:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5395cdb5b6b738bd8c01cc70000de4f86ae55b60/src/persistence/profile.h#L134)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Same as above
</blockquote></details>

Done.
___
*[src/persistence/profile.cpp, line 179 at r4](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsOwD0I5BZpnP58Ff:-Lk-r0LT47b8eaY6VK4E:bgk5yyw) ([raw file](https://github.com/qtox/qtox/blob/897f49f850abbd9d8eb536be2d14ee05b971cb02/src/persistence/profile.cpp#L179)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Return inplace
</blockquote></details>

Impossible without increasing the scope of the PR (initCore/loadDatabase call)
___
*[src/persistence/profile.cpp, line 304 at r5](https://reviewable.io/reviews/qtox/qtox/5729#-LiSsxqe1f-0FI4OxeOW:-Lk-rEJ28GwaO58IawGJ:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/5395cdb5b6b738bd8c01cc70000de4f86ae55b60/src/persistence/profile.cpp#L304)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

`goto end`

It can lead to problems if someone will add logic somewhere between `if`s.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sudden6 commented:
@Diadlo ping

Diadlo commented on src/persistence/profile.cpp:
```diff
@@ -120,7 +290,7 @@ Profile::Profile(QString name, const QString& password, bool isNewProfile,
  *
  * @example If the profile is already in use return nullptr.
```
Out of the scope. But maybe replace `@example` on `@note`?

sudden6 reviewed (changes requested):
> Reviewed 3 of 4 files at r12, 2 of 2 files at r13.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5729)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/profile.cpp, line 114 at r14](https://reviewable.io/reviews/qtox/qtox/5729#-LlT0JGiEYS6qh865eKC:-LlT0JGj3L8HCdiEvwhu:b-lccked) ([raw file](https://github.com/qtox/qtox/blob/007e062079eddba9758fada2fe2ee279c95bee49/src/persistence/profile.cpp#L114)):*
> ```cpp
>         if (data.isEmpty()) {
>             error = LoadToxDataError::DECRYPTION_FAILED;
>             goto end;
> ```

If we exit here the result is now `tmpKey` where before it was `nullptr`
___
*[src/persistence/profile.cpp, line 139 at r14](https://reviewable.io/reviews/qtox/qtox/5729#-LlT0dFB69JReK8iq6_u:-LlT0dFB69JReK8iq6_v:butee2) ([raw file](https://github.com/qtox/qtox/blob/007e062079eddba9758fada2fe2ee279c95bee49/src/persistence/profile.cpp#L139)):*
> ```cpp
>         if (!newKey) {
>             error = CreateToxDataError::COULD_NOT_DERIVE_KEY;
>             goto end;
> ```

before the return value was `nullptr` now it's `newKey` for all cases, I don't think that's intended.


<!-- Sent from Reviewable.io -->

Kribylet reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5729)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/profile.cpp, line 114 at r14](https://reviewable.io/reviews/qtox/qtox/5729#-LlT0JGiEYS6qh865eKC:-LlTHr9JCgX71W9KzSoF:b360mvj) ([raw file](https://github.com/qtox/qtox/blob/007e062079eddba9758fada2fe2ee279c95bee49/src/persistence/profile.cpp#L114)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

If we exit here the result is now `tmpKey` where before it was `nullptr`
</blockquote></details>

Fixed. Sharp catches!
___
*[src/persistence/profile.cpp, line 139 at r14](https://reviewable.io/reviews/qtox/qtox/5729#-LlT0dFB69JReK8iq6_u:-LlTHmP-D_sDi4cZhrlJ:bt2vr1b) ([raw file](https://github.com/qtox/qtox/blob/007e062079eddba9758fada2fe2ee279c95bee49/src/persistence/profile.cpp#L139)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

before the return value was `nullptr` now it's `newKey` for all cases, I don't think that's intended.
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r21.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5729#-:-LlTImlW4exozJPVhxcJ:b-jbsf8z)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->

Diadlo reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 6 of 6 files at r15, 2 of 2 files at r16, 4 of 4 files at r17, 1 of 1 files at r18, 4 of 4 files at r19, 2 of 2 files at r20.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5729#-:-LlqMM8wDTrautZar0O0:b-jbsf8z)* status: :shipit: complete! 2 of 1 LGTMs obtained

___
*[src/persistence/profile.cpp, line 200 at r18](https://reviewable.io/reviews/qtox/qtox/5729#-LlqLwNZ0mi2n8_nGrn4:-LlqLwNZ0mi2n8_nGrn5:b-ya81ng) ([raw file](https://github.com/qtox/qtox/blob/43cce7ed878549827e090cb46aa4bf6fa790b61d/src/persistence/profile.cpp#L200)):*
> ```cpp
>     switch (error) {
>     case CreateToxDataError::OK:
>         break;
> ```

Same as above
___
*[src/persistence/profile.cpp, line 168 at r21](https://reviewable.io/reviews/qtox/qtox/5729#-LlqMI1_5EE9PllsEnMH:-LlqMI1_5EE9PllsEnMI:b-mb5gsn) ([raw file](https://github.com/qtox/qtox/blob/01b45714a21ab13b68a62462b5074bdc58450981/src/persistence/profile.cpp#L168)):*
> ```cpp
>     switch (error) {
>     case LoadToxDataError::OK:
>         break;
> ```

Maybe `return true` here and `return false` in the end?


<!-- Sent from Reviewable.io -->

Kribylet reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5729)* status: :shipit: complete! 2 of 1 LGTMs obtained

___
*[src/persistence/profile.cpp, line 291 at r14](https://reviewable.io/reviews/qtox/qtox/5729#-LiSLEAZ0yuEKWJMrszJ-r14-291:-Lls5YniB_FIhTIv6rHj:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/007e062079eddba9758fada2fe2ee279c95bee49/src/persistence/profile.cpp#L291)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Out of the scope. But maybe replace `@example` on `@note`?
</blockquote></details>

Done.
___
*[src/persistence/profile.cpp, line 200 at r18](https://reviewable.io/reviews/qtox/qtox/5729#-LlqLwNZ0mi2n8_nGrn4:-Lls5ZPc6tpHiAh5Oxxs:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/43cce7ed878549827e090cb46aa4bf6fa790b61d/src/persistence/profile.cpp#L200)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Same as above
</blockquote></details>

Done.
___
*[src/persistence/profile.cpp, line 168 at r21](https://reviewable.io/reviews/qtox/qtox/5729#-LlqMI1_5EE9PllsEnMH:-Lls5Zik5rMSjmWJmUlC:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/01b45714a21ab13b68a62462b5074bdc58450981/src/persistence/profile.cpp#L168)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Maybe `return true` here and `return false` in the end?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
